### ISSUE TITLE
Uninstalling VS2017 Preview removed F#4.1

### ISSUE THREAD

**Issue by isaacabraham** [Microsoft/visualfsharp#5513]
## TLDR
It *appears* as though uninstalling VS2017 Preview removed the F#4.1 SDK from `Program Files (x86)`, even though Ionide use it and VS2017 "full" was installed as well. This shouldn't happen.

## Long, boring details
I had both VS2017 *and* VS2017 Preview installed. This morning I upgraded VS2017 to 15.8 and then uninstalled VS2017 Preview .

Afterwards, I noticed that Ionide was using F#4.0, and not F#4.1, for intellisense / highlighting. For example, `Option.defaultValue` was showing as an error. Even after I set up Ionide to use the latest F#4.5 for the FSI (which worked correctly), this didn't help the IDE code editor itself. After speaking with @Krzysztof-Cieslak, I learned that the IDE part uses FSAC, which in turn looks for F# tooling in `C:\Program Files (x86)\Microsoft SDKs\F#`. Lo and behold, this folder only had F#4.0 in it - there was no F#4.1.

I ran a repair on VS2017, but after the one hour download and reinstall (and losing settings...), it didn't help.

I then tried installing the Microsoft Build Tools 2017. All this did was install a *second* copy of VS2017, but still no F#4.1 folder.

Eventually I found the [F# Compiler SDK](http://download.microsoft.com/download/F/3/D/F3D6045E-4040-4058-ADAD-2698F1793CBC/Microsoft.FSharp.SDK.Core.msi) which was what I needed. I installed this and Ionide started working again.

## Repro
Unfortunately, I'm not going to be able to reproduce this. I can't even know for sure if it's the case - but it might be worth someone doing some investigation to see if this is a plausible chain of events before other people fall foul of it.

**Comment by cartermp** [comment]
cc @brettfo

**Comment by KevinRansom** [comment]
It should be noted that the build tools Sku is teeny tiny compared to VS.

**Comment by KevinRansom** [comment]
@isaacabraham  Build Tools sku can be downloaded from here:  https://visualstudio.microsoft.com/downloads/

Scroll down the page and select Tools for Visual Studio.

![image](https://user-images.githubusercontent.com/5175830/44172139-1eaa9d00-a091-11e8-8f89-75240a874d59.png)

**Comment by Krzysztof-Cieslak** [comment]
> The other points you make speak to larger issues around FSAC and so on - the maintainers of those projects should be brought into the loop if you think that they're doing the wrong thing. 

We accept PRs to the FSAC... even from MSFT employees ;]

**Comment by cartermp** [comment]
@isaacabraham what you're saying here:

> I installed VS2017. Then I installed VS 2017 Preview. Both install side-by-side. I then uninstalled Preview but didn't uninstall VS2017 "full". I was left without the SDK on my machine.

Is exactly the problem @KevinRansom is saying is fixed. Because there was no such thing as side-by-side for F# prior to 15.8 when installed by VS. VS has, for the longest time, installed a compiler SDK into a global location and marked it as, "I installed this here". That marking was done because one of the things people find most annoying about VS is that some components need to be separately uninstalled/deleted, leaving the impression that VS "dirtied" their machine.

The problem with having that global location was that other installs could simply overwrite it silently, causing any tools to automatically be opted in to a compiler SDK version that they likely did not test against. All of the following:

* Older VS versions
* FSAC
* Rider
* Type Providers
* Custom tools

Would have their most fundamental dependency changed out from under them with no traceability if the user upgraded Visual Studio. And had the tools been versioned according to their intent (i.e., make it clear that it's a VS component), then it's just putting VS-specific components in a non-VS folder.

Sorry this is affecting you, but unfortunately I don't think it's particularly actionable in terms of what we can do to change F# setup. VS will now install F# truly side-by-side, including the previews. I believe the proper recourse for FSAC is to probe:

`C:\Program Files (x86)\Microsoft Visual Studio\2017\<sku>\Common7\IDE\CommonExtensions\Microsoft\FSharp`

Where `sku` is either `Community`, `Professional`, or `Enterprise` (there is an express SKU of 2017 but it's up to you if you want to probe for that).

As an aside, once FSI for .NET Core is completely it might be reasonable for Ionide's fork of FSAC to just dump the dependency on the F# desktop compiler altogether, since .NET Core greatly simplifies things and was designed for these sorts of scenarios.

**Comment by isaacabraham** [comment]
@cartermp Thanks for explaining it a bit more for me :-) No need to apologies - I've worked around this problem, but yes, this is something that FSAC needs to be aware of (whomever maintains that. I have no clue).

Not sure whether they'd want to couple FSAC to `\2017\` though! Are we also saying that we'll always need the VS Build Tools (and an inherent coupling to VS for the FSAC) or something else?

**Comment by forki** [comment]
/cc @enricosada @Krzysztof-Cieslak

Isaac Abraham <notifications@github.com> schrieb am Do., 16. Aug. 2018,
16:51:

**Comment by baronfel** [comment]
I'm working on the change now. Would MSBuild logic also need updating to new paths?

```
  let msbuild =
#if SCRIPT_REFS_FROM_MSBUILD
      if not(RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) then
        //well, depends on mono version, but like this is mono >= 5.2 (msbuild on mono 5.0 sort of works)
        "msbuild"
#else
      if Utils.runningOnMono then "xbuild" // mono <= 5.0
#endif
      else
        let MSBuildPath =
            (programFilesX86 </> @"\MSBuild\14.0\Bin") + ";" +
            (programFilesX86 </> @"\MSBuild\12.0\Bin") + ";" +
            (programFilesX86 </> @"\MSBuild\12.0\Bin\amd64") + ";" +
            @"c:\Windows\Microsoft.NET\Framework\v4.0.30319\;" +
            @"c:\Windows\Microsoft.NET\Framework\v4.0.30128\;" +
            @"c:\Windows\Microsoft.NET\Framework\v3.5\"
        let ev = Environment.GetEnvironmentVariable "MSBuild"
        if not (String.IsNullOrEmpty ev) then ev
        else findPath MSBuildPath "MSBuild.exe"
```

**Comment by cartermp** [comment]
Unfortunately there isn't a way around this, but 2017 is probably a good place for a while. We won't be releasing VS 2019 for some time, and it's not yet been determined if there will be a new F# language version going with that version either (though I'd certainly like to ship one). And it's probably a good idea to keep probing for VS 2017 until such a time that F# 4.5 is considered "old".

**Comment by cartermp** [comment]
@baronfel Unfortunately I'm not sure, but if you're on Windows you can launch the VS Command prompt and type `set` to see locations for everything, including the MSBuild installed by VS.

**Comment by matthid** [comment]
Can we not bundle our own and don’t have a dependency at all?

**Comment by baronfel** [comment]
And/or make FSAC configurable with regards to which fsharpi/fsharpc/msbuild it uses.

**Comment by Krzysztof-Cieslak** [comment]
@isaacabraham

> whomever maintains that. I have no clue

😘

—-

@cartermp  @KevinRansom

How this all work with non-VS Windows F# installations? Is it actually possible to install F# on Windows without VS at all?

**Comment by cartermp** [comment]
Yes, the VS Build tools (mentioned earlier) are the recommended way to get the desktop fsc/fsi/etc. and all its required dependencies without installing VS.

**Comment by KevinRansom** [comment]
@Krzysztof-Cieslak Xamarin handle VS MAC deployment.   And fsharp/fsharp other non-windows platforms.

dotnet cli is of course our primary mechanism for non-windows deployment, which We handle with the dotnet cli team.

**Comment by Krzysztof-Cieslak** [comment]
@baronfel

> And/or make FSAC configurable with regards to which fsharpi/fsharpc/msbuild it uses.

This is doable with environmental variables - `MsBuild` for MsBuild path, `FsharpInstallationDir` (?) for fsharpc. FSAC doesn’t really depends on FSI - but for Ionide there is setting pointing to FSI path

**Comment by KevinRansom** [comment]
Hmm!!!,

why does it require the old msbuild?

**Comment by Krzysztof-Cieslak** [comment]
@KevinRansom

From quick glance at the code, this line looks problematic - https://github.com/Microsoft/visualfsharp/blob/master/fcs/FSharp.Compiler.Service.ProjectCrackerTool/ProjectCrackerTool.fs#L52

But, I have no idea whether it’s enough to just fix this line or there are some other assumptions in there. TBF, no one wants to touch project cracker any more - it’s been a source of constant problems for ever, and Enrico has created totally new tool for SDK based projects that’s working better. We hope that people will just slowly move to new format ;-)

**Comment by KevinRansom** [comment]
@Krzysztof-Cieslak --- It seems to use that for probing to find an msbuild,   sigh!!!!!!  --- so the same problem.

**Comment by cartermp** [comment]
It sounds like we have a resolution on getting Ionide updated with the correct probing logic for F# and MSBuild 15 (thanks @baronfel for putting together the PR and asking clarifying questions in slack about things!)

The VS uninstall behavior is not something that I think we'll fix, nor is it something we could fix except for servicing the 15.0.x release, because only 15.0.x and 15.8 are supported now that 15.8 is out. This has forever been the behavior (at least with VS 2017), and since we're moving forward as side-by-side, quirks with installing to a global location aren't a problem moving forward.

### Quick writeup

For posterity, the likely way this could have affected you today is:

1. Installed release 15.7.x or lower, which placed a stable F# 4.1 in the global location.
2. Installed a 15.8 preview prior to preview 5, which placed _some_ compiler bits that happened to be named F# 4.1 in the global location and overwrote the stable bits.
3. Upgraded the stable VS build to 15.8, which brought F# 4.5 along with it and installed side by side.
4. Uninstalled the preview, which had marked the global F# install as something it placed there, so it removed it.

The end result being that the F# 4.1 that was placed there by preview VS was removed when uninstalling preview VS, thus resulting in no F# 4.1 for Ionide to pick up.

If people had only installed VS preview 5, then they likely would not have had this issue, because that's when it brought a compiler fully side-by-side.

At least in testing on my own machine, upgrading from 15.7.x to 15.8 did not uninstall the global F# 4.1 location. This is because (a) the compiler is side by side, and (b) the way upgrades work is that only components that are registered as new are downloaded an installed. A full uninstall of VS is _not_ performed when upgrading.

Again, apologies @isaacabraham and others who run into this. I'll close this out, but please do continue to ask questions if you have them.
